### PR TITLE
Fix XML ElWriter bug allowing incomplete XML to be written

### PR DESCRIPTION
**Pull request by obi1kenobi**
## Motivation and Context
The XML `ElWriter` struct may currently be misused in a way that results in incomplete and invalid XML being written to the `&mut String` it contains. The root cause of the bug is that a user may forget to call `finish()` on `ElWriter`, and in doing so fail to write the closing `>` of the element represented by that `ElWriter`.

## Description
This PR contains two test cases that demonstrate the bug, as well as a fix: a `Drop` impl for `ElWriter` and a change to how it stores the mutable buffer, in order to avoid a lifetime problem between `drop()` and the move of the buffer into a `ScopeWriter`.

This approach is preferable over an alternative like `#[must_use]` on `ElWriter` and `ElWriter::write_attribute`, since that would only produce a warning lint instead of avoiding the problem entirely and guaranteeing the desired compile-time safety described in the docs.

## Testing
Added two test cases that highlight the problem. Ran `cargo test` inside `smithy-xml` and had everything succeed.

## Checklist
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated `aws/SDK_CHANGELOG.md` if applicable

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by obi1kenobi** [comment]
Thanks to @rcoh for explaining the idea behind the XML module, as well as the helpful discussion around verifying this is a bug and coming up with a fix!

File: rust-runtime/smithy-xml/src/encode.rs [hunk 1]
```diff
@@ -60,42 +60,55 @@ impl<'a> XmlWriter<'a> {
...
 
 pub struct ElWriter<'a, 'b> {
     start: &'b str,
-    doc: &'a mut String,
+    doc: Option<&'a mut String>,
 }
 
 impl<'a, 'b> ElWriter<'a, 'b> {
     pub fn write_attribute(&mut self, key: &str, value: &str) -> &mut Self {
-        write!(self.doc, " {}=\"{}\"", key, escape(value)).unwrap();
```
Comment (rcoh) [inline rust-runtime/smithy-xml/src/encode.rs:76]: can you add a private function along the lines of below that has a comment explaining why this is an option, why it won't be `None`, and generally, what's going on?
```rust
fn doc(&mut self) -> &'a mut String {
  // ...
}
```

File: rust-runtime/smithy-xml/src/encode.rs [hunk 2]
```diff
@@ -60,42 +60,55 @@ impl<'a> XmlWriter<'a> {
...
+        Self::write_end(doc);
         ScopeWriter {
-            doc: self.doc,
+            doc,
             start: self.start,
         }
     }
 }
 
+impl Drop for ElWriter<'_, '_> {
```
Comment (rcoh) [inline rust-runtime/smithy-xml/src/encode.rs:104]: can you add a quick comment about why this won't double-close?

**Review by rcoh — COMMENTED** [review]
This is great, thanks! See a couple of small requests below.

**Comment by rcoh** [comment]
Also, although the `CHANGELOG` box is checked, I don't see a `CHANGELOG` update in the diff, you may need to push

**Comment by rcoh** [comment]
CI failure is from `cargo fmt`. We have a pre-commit hook that can do that for you if you run `pre-commit install`.

**Comment by obi1kenobi** [comment]
Thanks for the suggestions! I'm still getting up to speed with style and commenting norms in Rust, so let me know if the comments or anything else can be improved.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
 vNext (Month Day, Year)
 =======================
+- :bug: Fix an issue where `smithy-xml` may have generated invalid XML (smithy-xml#719)
```
Comment (rcoh) [inline CHANGELOG.md:3]:
```suggestion
- :bug: Fix an issue where `smithy-xml` may have generated invalid XML (smithy-rs#719)
```

**Comment by rcoh** [comment]
LGTM! Thanks @obi1kenobi

**Review by rcoh — APPROVED** [review]

**Pull request merged**